### PR TITLE
Smarter resource listing

### PR DESCRIPTION
**Pull request by chesleybrown**
# What

- After testing support for Scheduled/Cron job kinds we realized that if a cluster isn't running the latest version of Kubernetes AND doesn't has alpha resources enabled that these would cause a failure and deploys would not work
- To fix this (and to provide an overall improvement) we will now only get a list of resources that match the type of resource you are deploying
- This should help reduce query load on all clusters as well as we will no longer get resources we don't need during a deployment

# Testing

1. Must use docker toolbox (not native Docker for Mac)
1. Run `./test-functional`

### CODE REVIEW

File: src/lib/manifests.js [hunk 1]
```diff
@@ -183,8 +208,11 @@ class Manifests extends EventEmitter {
 				this.emit("warn", err);
 			});
 
-			return this.kubectl
-				.list(this.supportedTypes.join(","), this.options.selector)
+			// Load manifest files
+			this.load();
```
Comment (chesleybrown) [inline src/lib/manifests.js]: Okay, it's not super pretty but I tried to fix some of the sync code that exists in 3a0342e

There are still a few places where files are being read sync... but it will require more code changes to fix so will probably have to do it in a different PR another day.

**Comment by chesleybrown** [comment]
Functional tests are passing:

```
test_functional_1  | 
test_functional_1  | > kit-deployer@0.0.0-semantic-release test-functional /
test_functional_1  | > mocha test/functional --recursive
test_functional_1  | 
test_functional_1  | 
test_functional_1  | 
test_functional_1  |   Functional
test_functional_1  |     when deploying to example cluster
test_functional_1  |       and example cluster does not exist yet
        ✓ should deploy without error (4168ms)
test_functional_1  |       and when running same deploy again (no differences)
        ✓ should deploy nothing and send webhooks (3971ms)
test_functional_1  |     when deploying to example cluster
test_functional_1  |       and exceeding AVAILABLE_TIMEOUT
        ✓ should trigger a TimeoutError (2076ms)
test_functional_1  |     when deploying multiple deployments cluster
test_functional_1  |       and multi-deployments cluster does not exist yet
        ✓ should deploy without error (7173ms)
test_functional_1  |       and multi-deployments cluster has already been deployed
        ✓ should deploy nothing and send webhooks (3987ms)
test_functional_1  |     when deploying mix deployment and service cluster
test_functional_1  |       and mix-deployment-service cluster does not exist yet
        ✓ should deploy without error (7112ms)
test_functional_1  |       and mix-deployment-service cluster has already been deployed
        ✓ should deploy nothing and send webhooks (3903ms)
test_functional_1  |     when no unique namespaces and deploying to cluster
      ✓ should deploy without error (702ms)
test_functional_1  |     when deploying a single job to a cluster
      ✓ should deploy without error (4029ms)
test_functional_1  | 
test_functional_1  | 
test_functional_1  |   9 passing (37s)
test_functional_1  | 
kitdeployer_test_functional_1 exited with code 0
```

File: src/lib/manifests.js [hunk 2]
```diff
@@ -90,28 +90,35 @@ class Manifests extends EventEmitter {
...
-					content: content
+			// validate directory, if it doesnt exist, skip processing.
+			//
+			glob(path.join(this.options.dir, this.options.cluster.metadata.name + "/**/*.yaml"), (err, files) => {
+				if (err) {
+					return reject(err);
+				}
+				_.each(files, (file) => {
+					// only add file if it matches selector
+					const content = yaml.safeLoad(fs.readFileSync(file, "utf8"));
```
Comment (c1freitas) [inline src/lib/manifests.js]: Can you do an `async` version of reading the file?

Comment (c1freitas) [inline src/lib/manifests.js]: Can you do an `async` version of reading the file?

File: src/lib/namespaces.js [hunk 3]
```diff
@@ -18,27 +18,31 @@ class Namespaces extends EventEmitter {
...
+				return namespaces;
+			}
+			glob(path.join(this.options.dir, this.options.clusterName + "/**/*.yaml"), (err, files) => {
+				if (err) {
+					return reject(err);
+				}
+				_.each(files, (file) => {
+					namespaces.push({
+						path: file,
+						content: yaml.safeLoad(fs.readFileSync(file, "utf8"))
```
Comment (c1freitas) [inline src/lib/namespaces.js]: Can you do an `async` version of reading the file?

File: src/lib/deployer.js [hunk 4]
```diff
@@ -55,137 +55,141 @@ class Deployer extends EventEmitter {
...
+					});
 				}
 
-				var kubectl = new Kubectl({
-					kubeconfig: configFile,
-					version: self.options.apiVersion
-				});
+				_.each(configFiles, function(configFile) {
+					// Parse the cluster yaml file to JSON
+					var config = yaml.safeLoad(fs.readFileSync(configFile, "utf8"));
```
Comment (c1freitas) [inline src/lib/deployer.js]: Can you do an `async` version of reading the file?

File: src/lib/manifests.js (hunk 2 above)
Comment (chesleybrown) [inline src/lib/manifests.js]: I _could_ but it's a bit more difficult here as these calls are wrapped in a sync loop (_.each) and after the loop runs it expects to have all the content. So in order to fix we need to instead have the loop create an array of promises that we want for all to resolve before continuing. It's all certainly possible, but required a few more code changes then I felt like making as it was really related to the changes/fixes I was originally trying to make.

Will take another look tomorrow morning and see if it's worth adding the change in on this PR.

Comment (c1freitas) [inline src/lib/manifests.js]: Got it. Looks good for now, definitely an improvement.

**Comment by chesleybrown** [comment]
@c1freitas I fixed the remaining sync calls. 😄 

Verified all functional tests are still passing after the changes.

```
test_functional_1  | 
test_functional_1  | > kit-deployer@0.0.0-semantic-release test-functional /
test_functional_1  | > mocha test/functional --recursive
test_functional_1  | 
test_functional_1  | 
test_functional_1  | 
test_functional_1  |   Functional
test_functional_1  |     when deploying to example cluster
test_functional_1  |       and example cluster does not exist yet
        ✓ should deploy without error (4102ms)
test_functional_1  |       and when running same deploy again (no differences)
        ✓ should deploy nothing and send webhooks (3909ms)
test_functional_1  |     when deploying to example cluster
test_functional_1  |       and exceeding AVAILABLE_TIMEOUT
        ✓ should trigger a TimeoutError (2025ms)
test_functional_1  |     when deploying multiple deployments cluster
test_functional_1  |       and multi-deployments cluster does not exist yet
        ✓ should deploy without error (7219ms)
test_functional_1  |       and multi-deployments cluster has already been deployed
        ✓ should deploy nothing and send webhooks (3947ms)
test_functional_1  |     when deploying mix deployment and service cluster
test_functional_1  |       and mix-deployment-service cluster does not exist yet
        ✓ should deploy without error (7178ms)
test_functional_1  |       and mix-deployment-service cluster has already been deployed
        ✓ should deploy nothing and send webhooks (3951ms)
test_functional_1  |     when no unique namespaces and deploying to cluster
      ✓ should deploy without error (684ms)
test_functional_1  |     when deploying a single job to a cluster
      ✓ should deploy without error (4134ms)
test_functional_1  | 
test_functional_1  | 
test_functional_1  |   9 passing (37s)
test_functional_1  | 
kitdeployer_test_functional_1 exited with code 0
```
